### PR TITLE
TPM signature provider: use TPM for nodeos signing 📦

### PR DESCRIPTION
**Pull request by spoonincode**
## Change Description

Add a TPM signature provider to signature_provider_plugin allowing nodeos to sign with a TPM. This signature provider has a few limitations:
* It only operates with persistent keys stored in the owner hierarchy
* No additional authentication on the hierarchy is supported (for example if the hierarchy requires an additional password/PIN auth)
* For PCR based policies, which are supported, they can only be specified on the sha256 PCR bank

A new tool, `eosio-tpmtool`, is included that can create keys in the TPM that are usable by nodeos. By design it is unable to remove keys. If more flexibly is desired (such as importing keys in to the TPM), a user may use external tools.

Usage example:
Start up a TPM software simulator
```
swtpm socket -p 2222 --tpm2 --tpmstate dir=/tmp/tpmstate --ctrl type=tcp,port=2223 --flags startup-clear
```

Create a key
```
$ eosio-tpmtool -c -T swtpm:port=2222
PUB_R1_5cgfoaDAacuE6iEdJE1GjVfJ65ftGtgFS8ACNpHJPRbYCcuHMQ
```

Use the key as a signature provider in nodeos.
```
signature-provider = PUB_R1_5cgfoaDAacuE6iEdJE1GjVfJ65ftGtgFS8ACNpHJPRbYCcuHMQ=TPM:swtpm:port=2222
```

Create a key with a policy such that it can only be used if the given sha256 PCRs are the current value
```
$ eosio-tpmtool -c -T swtpm:port=2222 -p5 -p7
PUB_R1_5SnCFs9JzXCXQ1PivjqwygZzSc3Qu5jK5GXf8C3aYNManLz7zq
```

Use the key as a signature provider in nodes with the specified PCR policy. The policy is not saved anywhere, so you will need to specify it again here.
```
signature-provider = PUB_R1_5cgfoaDAacuE6iEdJE1GjVfJ65ftGtgFS8ACNpHJPRbYCcuHMQ=TPM:swtpm:port=2222|5,7
```

In addition to the unit tests which operate with a TPM simulator, this code was tested manually on a handful of real TPMs.

CI integration notes:
TPM support is optional.. This PR doesn’t add integration in to the docker build, build scripts, nor CI, so support isn't enabled in any of these builds. There are a handful of problems integrating support with these various platforms. I've waffled on this quite a bit and don't have a good approach. Boxed will likely be using a separate build pipeline in the short term, so maybe the build changes should be solely integrated there for now? Not sure..

## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [X] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/tpm-helpers/include/eosio/tpm-helpers/tpm-helpers.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+private:
+   struct impl;
+   constexpr static size_t fwd_size = 256;
+   fc::fwd<impl,fwd_size> my;
+};
+
+boost::container::flat_set<fc::crypto::public_key> get_all_persistent_keys(const std::string& tcti);
+fc::crypto::public_key create_key(const std::string& tcti, const std::vector<unsigned>& pcrs);
+
+class nv_data {
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/tpm-helpers/include/eosio/tpm-helpers/tpm-helpers.hpp:29]: it's used for something in the platform-components repo

File: libraries/tpm-helpers/tpm-helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+
+   static constexpr unsigned create_size = 2048;
+};
+
+nv_data::nv_data(const std::string& tcti, unsigned nv_index, const std::vector<unsigned>& pcrs) : my(tcti, pcrs) {
+   TSS2_RC rc;
+
+   rc = Esys_TR_FromTPMPublic(my->esys_ctx.ctx(), TPM2_HR_NV_INDEX + nv_index, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &my->nv_handle);
+   if(rc) {
+      //let's try to create it...
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/tpm-helpers/tpm-helpers.cpp:398]: Yes quite likely. What will ultimately help protect against those sort of abuse vectors is getting attestment in place via TPM's certify and/or quote operations. That'll have to wait until later though.

File: plugins/signature_provider_plugin/signature_provider_plugin.cpp [hunk 3]
```diff
@@ -41,6 +44,28 @@ class signature_provider_plugin_impl {
...
 #endif
 
+#ifdef ENABLE_TPM
+      signature_provider_plugin::signature_provider_type
+      make_tpm_signature_provider(const std::string& spec_data, const chain::public_key_type pubkey) const {
+         std::vector<string> params;
+         std::vector<unsigned> pcrs;
+
+         boost::split(params,spec_data,boost::is_any_of("|"));
+         if(params.size() >= 2) {
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply plugins/signature_provider_plugin/signature_provider_plugin.cpp:54]: nah. Are you suggesting to throw an error if there are more pipes, or??

File: libraries/tpm-helpers/include/eosio/tpm-helpers/tpm-helpers.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#include <boost/container/flat_set.hpp>
+
+#include <set>
+
+namespace eosio::tpm {
+
+class tpm_key {
+public:
+   tpm_key(const std::string& tcti, const fc::crypto::public_key& pubkey, const std::vector<unsigned>& pcrs);
+   ~tpm_key();
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/tpm-helpers/include/eosio/tpm-helpers/tpm-helpers.hpp:16]: I think fc::fwd will stop all those from being copied/assigned. But yeah it's probably worth being more explicit about it.

[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply libraries/tpm-helpers/include/eosio/tpm-helpers/tpm-helpers.hpp:16]: `fc::fwd` allows copy operations.  It looks to me as though attempting to copy them will cause a link error, except in `tmp-helpers.cpp`.

File: plugins/signature_provider_plugin/signature_provider_plugin.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply plugins/signature_provider_plugin/signature_provider_plugin.cpp:54]: An error

File: libraries/tpm-helpers/tpm-helpers.cpp [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+      TPM2B_PUBLIC* pub = nullptr;
+      auto cleanup_pub = fc::make_scoped_exit([&]() {free(pub);});
+
+      rc = Esys_ReadPublic(esys_ctx.ctx(), object, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &pub, NULL, NULL);
+      if(rc)
+         continue;
+      if((pub->publicArea.objectAttributes & TPMA_OBJECT_SIGN_ENCRYPT) == 0)
+         continue;
+      if(pub->publicArea.objectAttributes & TPMA_OBJECT_RESTRICTED)
+         continue;
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/tpm-helpers/tpm-helpers.cpp:237]: agreed, but tpm2-tss is missing a #define for that :disappointed:  I guess too new. will think about making a define locally..

**Review by swatanabe-b1 — APPROVED** [review]

**Pull request merged**